xmtp_mls/src/groups/key_package_cleaner_worker.rs

```diff
@@ -0,0 +1,159 @@
+                        installation_id,
+                        "Pool disconnected. task will restart on reconnect"
+                    );
+                    break;
+                } else {
+                    tracing::error!(inbox_id, installation_id, "sync worker error {err}");
+                    // Wait 2 seconds before restarting.
+                    xmtp_common::time::sleep(WORKER_RESTART_DELAY).await;
+                }
+            }
+        });
+    }
+}
+
+impl<ApiClient, Db> KeyPackagesCleanerWorker<ApiClient, Db>
+where
+    ApiClient: XmtpApi + Send + Sync + 'static,
+    Db: XmtpDb + Send + Sync + 'static,
+{
+    /// Delete a key package from the local database.
+    pub(crate) fn delete_key_package(
+        &self,
+        hash_ref: Vec<u8>,
+    ) -> Result<(), IdentityError> {
+        let openmls_hash_ref = crate::identity::deserialize_key_package_hash_ref(&hash_ref)?;
+        tracing::info!("key store");
+        self.client.mls_provider().key_store().delete_key_package(&openmls_hash_ref)?;
+
+        Ok(())
+    }
+    
+    /// Delete all the expired keys
+    async fn delete_expired_key_packages(&mut self) -> Result<(), KeyPackagesCleanerError> {
+        let provider = self.client.mls_provider();
+        let conn = provider.db();
+
+        match conn.get_expired_key_packages() {
+            Ok(expired_kps) if !expired_kps.is_empty() => {
+                // Delete from local db
+                for kp in &expired_kps {
+                    if let Err(err) = self.delete_key_package(kp.key_package_hash_ref.clone()) {
+                        tracing::error!("Couldn't delete KeyPackage: {:?}", err);
+                    }
+                }
+
+                // Delete from database using the max expired ID
+                if let Some(max_id) = expired_kps.iter().map(|kp| kp.id).max() {
+                    conn.delete_key_package_history_entries_before_id(max_id)?;
+                    tracing::info!(
+                    "Deleted {} expired key packages (up to ID {}) from local DB and state",
+                    expired_kps.len(),
+                    max_id
+                );
+                }
+            }
+            Ok(_) => {
+                tracing::debug!("No expired key packages to delete");
+            }
+            Err(e) => {
+                tracing::error!("Failed to fetch expired key packages: {:?}", e);
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Check if we need to rotate the keys and upload new keypackage if the las one rotate in has passed
+    async fn rotate_last_key_package_if_needed(&mut self) -> Result<(), KeyPackagesCleanerError> {
+        self.client.rotate_and_upload_key_package(true).await?;
+        Ok(())
+    }
+    
+    async fn run(&mut self) -> Result<(), KeyPackagesCleanerError> {
+        let mut intervals = xmtp_common::time::interval_stream(INTERVAL_DURATION);
+        while (intervals.next().await).is_some() {
```

Comment:
Can we slow the interval to 5s or maybe 10s?

Reply:
A window of 5-10s should be fine, yeah? If the user closes the app and comes back in 30 minutes, the kp won't be deleted for 30 minutes.